### ISSUE TITLE
JPA Repository unnecessarily requires transactions

### ISSUE THREAD

**Issue by l-dobrev** [spring-projects/spring-data-jpa#2226]
TLDR; The `SimpleJpaRepository` base is marked `@Transactional` which requests a transaction for every operation, even when there is no need for that.

Applying `@Transactional(propagation = Propagation.SUPPORTS, readOnly = true)` to a service method is a way to request a read-only database operation without creating a dedicated transaction.
The use-case in mind is using PostgreSQL master-slave replication with a PgPool infront of the cluster. The pool will always create a connection to the master server if a transaction is requested (required for isolation and locking to work). Whenever a read-only process is done, not creating a wrapping transaction allows the queries to be forwarded to slave servers thus reducing the stress on the master server.

The current blank `@Transactional` annotation on the `SimpleJpaRepository` breaks this in a way that may be hard to replicate. Transactional services (Supports+Read-Only) that load entities via the JPA Repositories will have their JPA Context paused, a new nested JPA Context is created, a new Transaction is created just for the JPA method, the entity/entities are loaded within the nested context and returned. The nested context and transaction are closed, and the entities end up being used outside their native JPA Context, which ends up crashing whenever the entity's relations (lazy ones) are requested.
These exceptions are very unexpected as normally a "Session is closed" exception is thrown while an active session is available.

The reason for this is that the blank `@Transactional` annotation has more specificity than a `@Transactional(propagation = Propagation.SUPPORTS, readOnly = true)` which is on the outside.

The side effects range from unexpected "Session is closed" exceptions to situations where multiple (sometimes hundreds) of transactions are repeatedly created and committed for a single service call.

**Comment by l-dobrev** [comment]
Correction: the `SimpleJpaRepository` is marked `@Transactional(readOnly = true)` no a blank `@Transactional`.

**Comment by schauder** [comment]
I understand your concern regarding PostgreSQLs master-slave setup. But that does not warrant a general change in behaviour I think.

I do not understand the argument about lazy loading failing when a transaction is required, because it should fail if non is present in the first place. From the JPA spec: 

> The find method (provided it is invoked without a lock or invoked with LockModeType.NONE) and the getReference method are not required to be invoked within a transaction. If an entity man- ager with transaction-scoped persistence context is in use, the resulting entities will be detached; if an entity manager with an extended persistence context is used, they will be managed.

**Comment by schauder** [comment]
I understand that, but you example should also crash without the `@Transactional` annotation on `SimpleJpaRepository`.

**Comment by l-dobrev** [comment]
@schauder it does **not** crash with the `@Transactional(propagation = SUPPORTS, readOnly = true)` which is the suggested resolution.

**Comment by schauder** [comment]
I would need a reproducer for that. Possibly as part of your PR.

**Comment by schauder** [comment]
Thanks, I'll take a look.

**Comment by schauder** [comment]
Now that is _interesting_ behaviour. I must admit I was quite surprised to see that lazy loading in deed did work. My expectation was that it wouldn't since there is no transaction/session and therefore nothing for lazy loading to work with.

Here is what is happening, if I got everything right.
You have two levels of methods: Service methods which call repository methods.
The main service methods have `PROPAGATION_SUPPORTS` while the repository methods have effectively `PROPAGATION_SUPPORTS` with your PR and `PROPAGATION_REQUIRED` without your PR.

The trick is that `PROPAGATION_SUPPORTS` actually does create a transaction, basically by asking if one exists.
This is the behaviour of the `JpaTransactionManager`.
Both service and repository share the same transaction this way and neither bothers to close it, since they never asked for a new one.
Therefore the transaction and the `EntityManager` are still open when control exited the repository and lazy loading works.

Without the PR the repository method considers the transaction a new one and therefore closes it on exit and therefore lazy loading outside the repository fails.

This is what I actually consider the correct behaviour.

The opening of an `EntityManager` without closing it by the `JpaTransactionManager` for `PROPAGATION_SUPPORTED` sounds wrong to me, especially when you consider it in a web app, where the same `EntityManager` bound to the thread could easily end up in a different web request, resulting in all kinds of funny effects. I'll ping the Spring Framework team to take a look if this description is correct and if that is the expected behaviour.

For this reason we don't want `PROPAGATION_SUPPORTED` as a default on the repository.

In order to solve your original problem though you may overwrite the methods in question and apply a `@Transactional` annotation with the configuration you prefer.

**Comment by l-dobrev** [comment]
@schauder I can not necessarily agree with your analysis, because there is a difference in creating a JPA Entity Manager (at any `@Transactional`) and creating a JDBC Transaction (at `@Transactional(propagation = REQUIRED)`).
The `@Transactional(propagation = SUPPORTS)` is expected to create a JPA Entity Manager instance, but **not** a JDBC Transaction.

Also note that simply using the JPA Entity Manager inside a `@Transactional(propagation = SUPPORTS)` works as expected (lazy loading and all), but does **not** work when doing it via a JPA Repository.

**Comment by schauder** [comment]
What do you mean by "works as expected"? Because my expectation is still that lazy loading will fail, which seems to be different from your expectation.

**Comment by mp911de** [comment]
I agree that `propagation = REQUIRED` will start a JPA/JDBC (`entityManager.getTransaction().begin()`) transaction while `propagation = SUPPORTS` isn't starting calling `entityManager.getTransaction().begin()`. The other difference is that `propagation = SUPPORTS` binds resources to the current thread and these resources remain remain open which leads to a resource leak.

**Comment by l-dobrev** [comment]
@schauder using JPA Entity Manager in `propagation=SUPPORTS` loads entities that successfully navigate lazy relations.
```
em.find(Loop.class, Integer.valueOf(1)).getChildren().size(); // Does not crash
```
```
loopRepo.findById(Integer.valueOf(1)).getChildren().size(); // Crashes
```

@mp911de my experience shows, that resources are cleaned up when the `@Transactional(propagation = SUPPORTS)` method completes. At least it seems that the JPA Entity Manager is disposed properly. Did I miss something there?

**Comment by mp911de** [comment]
> At least it seems that the JPA Entity Manager is disposed properly. Did I miss something there?

The cleanup happens in your test fixture through `tearDown` as the called method is marked with `@Transactional(propagation=REQUIRED)`. That method uses the thread-bound entity manager and closes it. You can verify this in `JpaTransactionManager.doCleanupAfterCompletion(…)`.

**Comment by l-dobrev** [comment]
@mp911de I am not sure that I am looking at the right place, but **after** the `propagation=SUPPORTS` method has finished there is nothing left in the `TransactionSynchronizationManager.getResourceMap()`. Does that not mean, that any thread-bound resources have been released/closed?

**Comment by mp911de** [comment]
Let me have another look on Monday so I can get back to you with further details.

**Comment by schauder** [comment]
#1041 Seems similar/related. There we effectively start a transaction by using the `EntityManager` and nothing is there to close it.

**Comment by l-dobrev** [comment]
@schauder #1041 does seem related.
Though the original issue on SO seems to hint, that `@Transactional` on the service level did not help, while I would expect that to work.

**Comment by mp911de** [comment]
I've had a look at what happens within Spring Framework in combination with the `EntityManager` using `SUPPORTS` propagation mode. As part of the `@Transactional` method support, the outermost call to a `@Transactional(propagation=SUPPORTS)` method performs resource synchronization cleanup. The JPA transaction manager registers `TransactionalEntityManagerSynchronization` and cleaning up resource synchronization actually closes the bound `EntityManager`, so we do not end up with a resource leak.

Another aspect that plays into the change is the auto-commit mode with which the database is left. With explicit transactions, reads participate in transaction isolation. Using `SUPPORTS` propagation doesn't enable explicit transactions and therefore multiple reads are not guaranteed to be consistent. Consider eager loading of an entity graph that requires multiple `SELECT` queries as a possible example. Any concurrent data modifications to the database could interfere with loading data which renders an inconsistent result.

Therefore, we're going to keep the propagation mode as-is.

**Comment by l-dobrev** [comment]
If anyone happens to be looking for a work-around: avoid using Spring JPA Repositories when using `propagation=SUPPORTS` at the service level, and rely on Entity Manager directly to perform database operations.

**Comment by whching** [comment]
Just come across this case and would like to share my comments.
Regarding the followings,
> Using SUPPORTS propagation doesn't enable explicit transactions and therefore multiple reads are not guaranteed to be consistent. Consider eager loading of an entity graph that requires multiple `SELECT` queries as a possible example. Any concurrent data modifications to the database could interfere with loading data which renders an inconsistent result.

I believe even with propagation=REQUIRED (current implementation in SimpleJpaRepository), it is not guaranteed to have consistent read for most of the databases (e.g. Oracle). If I do need a consistent entity graph, IMHO, I should use FETCH JOIN. So, what I mean is, setting propagation=REQUIRED is not necessary a reason for consistency.
